Userland/Utilities/less.cpp

```diff
@@ -0,0 +1,248 @@
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <ctype.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/ioctl.h>
+#include <termios.h>
+#include <unistd.h>
+
+static struct termios g_save;
+static struct winsize g_wsize;
+
+static void setup_tty()
+{
+    // Save previous tty settings.
+    if (tcgetattr(STDOUT_FILENO, &g_save) == -1) {
+        perror("tcgetattr(3)");
+    }
+
+    // Get the window size.
+    if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &g_wsize) == -1) {
+        perror("ioctl(2)");
+    }
+
+    struct termios raw = g_save;
+    raw.c_lflag &= ~(ECHO | ICANON);
+
+    // Disable echo and line buffering
+    if (tcsetattr(STDOUT_FILENO, TCSAFLUSH, &raw) == -1) {
+        perror("tcsetattr(3)");
+    }
+
+    // Save cursor and switch to alternate buffer.
+    out("\e[s\e[?1047h");
+}
+
+static void teardown_tty()
+{
+    if (tcsetattr(STDOUT_FILENO, TCSAFLUSH, &g_save) == -1) {
+        perror("tcsetattr(3)");
+    }
+
+    out("\e[?1047l\e[u");
+}
+
+static Vector<String> wrap_line(const Utf8View& string, size_t width)
+{
+    Vector<String> lines;
+
+    StringBuilder builder;
+    size_t offset = 0;
+
+    bool in_ansi = false;
+    for (auto codepoint : string) {
+        if (offset >= width) {
+            builder.append('\n');
+            lines.append(builder.build());
+            builder.clear();
+            offset = 0;
+        }
+
+        builder.append(codepoint);
+
+        if (codepoint == '\e')
+            in_ansi = true;
+
+        if (!in_ansi)
+            // FIXME: calcuate the printed width of the character.
+            offset++;
+
+        if (isalpha(codepoint))
+            in_ansi = false;
+    }
+
+    if (builder.length() > 0)
+        lines.append(builder.build());
+
+    return lines;
+}
+
+class Pager {
+public:
+    Pager(FILE* file, FILE* tty, size_t width, size_t height)
+        : m_file(file)
+        , m_tty(tty)
+        , m_width(width)
+        , m_height(height)
+    {
+    }
+
+    void up()
+    {
+        up_n(1);
+    }
+
+    void down()
+    {
+        down_n(1);
+    }
+
+    void up_page()
+    {
+        up_n(m_height - 1);
+    }
+
+    void down_page()
+    {
+        down_n(m_height - 1);
+    }
+
+    void up_n(size_t n)
+    {
+        if (m_line == 0)
+            return;
+
+        m_line = (m_line > n) ? m_line - n : 0;
+
+        // Clear screen and reset cursor position.
+        out("\e[2J\e[0G\e[0d");
+        write_range(m_line, m_height - 1);
+        status_line();
+        fflush(m_tty);
+    }
+
+    void down_n(size_t n)
+    {
+        clear_status();
+
+        while (n - (m_lines.size() - m_line) > 0) {
+            if (!read_line())
+                break;
+        }
+        m_line += write_range(min(m_line + m_height - 1, m_line + (m_lines.size() - m_line)), n);
+        status_line();
+
+        fflush(m_tty);
+    }
+
+    void init()
+    {
+        while (m_lines.size() < m_height) {
+            read_line();
```

Comment:
What if we don't have enough lines to fill the screen?
won't this loop forever?

Reply:
yep

Reply:
I suppose that's intentional, it'll just die with sigpipe?

Reply:
no, I've fixed it. it should break when read_line() is false.